transaction/undoTx.go

```diff
@@ -154,43 +124,47 @@ func _initUndoTx(size int) *undoTx {
 }
 
 func NewUndoTx() TX {
-	if undoPool[0] == nil {
+	if undoPool == nil {
 		log.Fatal("Undo log not correctly initialized!")
 	}
-	t := <-undoPool[0]
-	return t
-}
-
-func NewLargeUndoTx() TX {
-	if undoPool[1] == nil {
-		log.Fatal("Undo log not correctly initialized!")
-	}
-	t := <-undoPool[1]
+	t := <-undoPool
 	return t
 }
 
 func releaseUndoTx(t *undoTx) {
 	t.abort()
-	if t.large {
-		undoPool[1] <- t
-	} else {
-		undoPool[0] <- t
-	}
+	undoPool <- t
 }
 
+// also takes care of increasing/decreasing the number of entries log can hold
 func (t *undoTx) updateLogTail(tail int) {
-	// atomic update
-	if t.large && tail >= LENTRYSIZE {
-		log.Fatal("Too large transaction. Already logged ",
-			LENTRYSIZE, " entries")
-	} else if !t.large && tail >= SENTRYSIZE {
-		log.Fatal("Too large transaction. Already logged ",
-			SENTRYSIZE, " entries")
+	if tail >= t.nEntry {
+		newE := 2 * t.nEntry // Double number of entries which can be stored
+		newLog := pmake([]entry, newE)
+		copy(newLog, t.log)
+		runtime.PersistRange(unsafe.Pointer(&newLog[0]),
+			uintptr(newE)*unsafe.Sizeof(newLog[0])) // Persist new log
+		oldLogSliceHdr := (*sliceHeader)(unsafe.Pointer(&t.log))
+		newLogSliceHdr := (*sliceHeader)(unsafe.Pointer(&newLog))
+
+		// Essentially doing t.log = newLog, but this needs to be atomic
+		// Each of the below three updates are atomic
+		// A crash in the middle should be fine
+		oldLogSliceHdr.data = newLogSliceHdr.data
+		oldLogSliceHdr.cap = newLogSliceHdr.cap
+		oldLogSliceHdr.len = newLogSliceHdr.len
+
+		t.nEntry = newE
+		runtime.PersistRange(unsafe.Pointer(t), unsafe.Sizeof(*t))
+	} else if tail == 0 { // reset to original size
+		t.log = t.log[:NUMENTRIES]
+		t.nEntry = NUMENTRIES
+		runtime.PersistRange(unsafe.Pointer(t), unsafe.Sizeof(*t))
 	}
-	runtime.Fence() //TODO: Check if this can be removed
+
+	// common case
 	t.tail = tail
```

Comment:
In the latest commit, I allocate a new log of entries when tail is set to 0, and if the log had grown before. So this will be fixed. Marking resolved for now.